[BUG] csv_test.py::test_csv_fallback FAILED on the EMR Cluster

**Describe the bug**
csv_test.py::test_csv_fallbac FAILED on the EMR Cluster
=================================== FAILURES ===================================
______ test_csv_fallback[spark.rapids.sql.format.csv.enabled-read_csv_df] ___

spark_tmp_path = '/tmp/pyspark_tests//818508/'
read_func = <function read_csv_df at 0x7f7b23de25f0>
disable_conf = 'spark.rapids.sql.format.csv.enabled'

    @allow_non_gpu('FileSourceScanExec')
    @pytest.mark.parametrize('read_func', [read_csv_df, read_csv_sql])
    @pytest.mark.parametrize('disable_conf', ['spark.rapids.sql.format.csv.enabled', 'spark.rapids.sql.format.csv.read.enabled'])
    def test_csv_fallback(spark_tmp_path, read_func, disable_conf):
        data_gens =[
            StringGen('(\\w| |\t|\ud720){0,10}', nullable=False),
            byte_gen, short_gen, int_gen, long_gen, boolean_gen, date_gen]
    
        gen_list = [('_c' + str(i), gen) for i, gen in enumerate(data_gens)]
        gen = StructGen(gen_list, nullable=False)
        data_path = spark_tmp_path + '/CSV_DATA'
        schema = gen.data_type
        updated_conf=_enable_all_types_conf.copy()
        updated_conf[disable_conf]='false'
    
        reader = read_func(data_path, schema)
        with_cpu_session(
                lambda spark : gen_df(spark, gen).write.csv(data_path))
        assert_gpu_fallback_collect(
                lambda spark : reader(spark).select(f.col('*'), f.col('_c2') + f.col('_c3')),
                'FileSourceScanExec',
>               conf=updated_conf)
integration_tests/src/main/python/csv_test.py[0m:292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
integration_tests/src/main/python/asserts.py[0m:302: in assert_gpu_fallback_collect
    conf=conf)
integration_tests/src/main/python/spark_session.py[0m:95: in with_gpu_session
    return with_spark_session(func, conf=copy)

integration_tests/src/main/python/spark_session.py:68: in with_spark_session
c(_spark)
integration_tests/src/main/python/asserts.py:179: in <lambda>
k = lambda spark: limit_func(spark).collect()
integration_tests/src/main/python/csv_test.py:290: in <lambda>
ark : reader(spark).select(f.col('*'), f.col('_c2') + f.col('_c3')),
integration_tests/src/main/python/csv_test.py:180: in read_impl
bug_df(reader.csv(data_path))
integration_tests/src/main/python/data_gen.py:674: in debug_df
LLECTED\n{}'.format(df.collect()))
/mnt/yarn/usercache/hadoop/appcache/application_1618918487651_0003/container_1618918487651_0003_01_000001/pyspark.zip/pyspark/sql/dataframe.py:597: in collect
 = self._jdf.collectToPython()
/mnt/yarn/usercache/hadoop/appcache/application_1618918487651_0003/container_1618918487651_0003_01_000001/py4j-0.10.9-src.zip/py4j/java_gateway.py:1305: in __call__
elf.gateway_client, self.target_id, self.name)
/mnt/yarn/usercache/hadoop/appcache/application_1618918487651_0003/container_1618918487651_0003_01_000001/pyspark.zip/pyspark/sql/utils.py:134: in deco
    raise_from(converted)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

e = IllegalArgumentException('Part of the plan is not columnar class org.apache.spark.sql.execution.datasources.v2.BatchSc...:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\n', None)

>   ???
E   pyspark.sql.utils.IllegalArgumentException: Part of the plan is not columnar class org.apache.spark.sql.execution.datasources.v2.BatchScanExec
E   BatchScan[_c0#220142, _c1#220143, _c2#220144, _c3#220145, _c4#220146L, _c5#220147, _c6#220148] CSVScan DataFilters: [], Location: InMemoryFileIndex[hdfs://ip-172-31-7-149.us-west-2.compute.internal:8020/tmp/pyspark_tests/818508..., PartitionFilters: [], ReadSchema: struct<_c0:string,_c1:tinyint,_c2:smallint,_c3:int,_c4:bigint,_c5:boolean,_c6:date>, PushedFilters: []

<string>:3: IllegalArgumentException
----------------------------- Captured stdout call -----------------------------
### CPU RUN ###
COLLECTED
[Row(_c0='z0N1e99PS', _c1=53, _c2=-29908, _c3=None, _c4=2451761049700751885, _c5=True, _c6=datetime.date(5699, 9, 5)), Row(_c0='QzvqbD05AF', _c1=-24, _c2=-24633, _c3=-204549548, _c4=3368872244882951150, _c5=True, _c6=datetime.date(4, 2, 29)), Row(_c0='ZqJYqy\tVOF', _c1=53, _c2=31152, _c3=-673166729, _c4=-4519395617304985747, _c5=None, _c6=datetime.date(2138, 12, 29)), Row(_c0='5GGb0KP4D_', _c1=101, _c2=-19589, _c3=709206688, _c4=-5600721832451008419, _c5=True, _c6=datetime.date(1308, 4, 5)), Row(_c0='vAK8rs2Hod', _c1=40, _c2=-2435, _c3=857391859, _c4=-3384071289586606412, _c5=True, _c6=datetime.date(2340, 4, 19)), Row(_c0='jhWh7aw_NU', _c1=-71, _c2=682, _c3=-26756290, ...


**Steps/Code to reproduce bug**
Run pytests on EMR cluster with the script https://github.com/NVIDIA/spark-rapids/blob/branch-0.5/integration_tests/run_pyspark_from_build.sh#L129~L131

**Expected behavior**
All the python tests PASS

**Environment details (please complete the following information)**
 - Environment location: EMR Yarn cluster

--- a/integration_tests/src/main/python/csv_test.py
+++ b/integration_tests/src/main/python/csv_test.py
@@ -268,21 +268,21 @@ def test_round_trip(spark_tmp_path, data_gen, v1_enabled_list):
             lambda spark : spark.read.schema(schema).csv(data_path),
             conf=updated_conf)
 
-@allow_non_gpu('FileSourceScanExec')
+@allow_non_gpu('FileSourceScanExec', 'BatchScanExec')
 @pytest.mark.parametrize('read_func', [read_csv_df, read_csv_sql])
 @pytest.mark.parametrize('disable_conf', ['spark.rapids.sql.format.csv.enabled', 'spark.rapids.sql.format.csv.read.enabled'])
 def test_csv_fallback(spark_tmp_path, read_func, disable_conf):
     data_gens =[
         StringGen('(\\w| |\t|\ud720){0,10}', nullable=False),
         byte_gen, short_gen, int_gen, long_gen, boolean_gen, date_gen]
- 
+
     gen_list = [('_c' + str(i), gen) for i, gen in enumerate(data_gens)]
     gen = StructGen(gen_list, nullable=False)
     data_path = spark_tmp_path + '/CSV_DATA'
     schema = gen.data_type
     updated_conf=_enable_all_types_conf.copy()
     updated_conf[disable_conf]='false'
- 
+
     reader = read_func(data_path, schema)
     with_cpu_session(
             lambda spark : gen_df(spark, gen).write.csv(data_path))
@@ -345,7 +345,7 @@ def test_ts_formats_round_trip(spark_tmp_path, date_format, ts_part, v1_enabled_
 
 @pytest.mark.parametrize('v1_enabled_list', ["", "csv"])
 def test_input_meta(spark_tmp_path, v1_enabled_list):
-    gen = StructGen([('a', long_gen), ('b', long_gen)], nullable=False) 
+    gen = StructGen([('a', long_gen), ('b', long_gen)], nullable=False)
     first_data_path = spark_tmp_path + '/CSV_DATA/key=0'
     with_cpu_session(
             lambda spark : gen_df(spark, gen).write.csv(first_data_path))

--- a/integration_tests/src/main/python/csv_test.py
+++ b/integration_tests/src/main/python/csv_test.py
@@ -268,7 +268,7 @@ def test_round_trip(spark_tmp_path, data_gen, v1_enabled_list):
             lambda spark : spark.read.schema(schema).csv(data_path),
             conf=updated_conf)
 
-@allow_non_gpu('FileSourceScanExec', 'BatchScanExec')
+@allow_non_gpu('org.apache.spark.sql.execution.LeafExecNode')
 @pytest.mark.parametrize('read_func', [read_csv_df, read_csv_sql])
 @pytest.mark.parametrize('disable_conf', ['spark.rapids.sql.format.csv.enabled', 'spark.rapids.sql.format.csv.read.enabled'])
 def test_csv_fallback(spark_tmp_path, read_func, disable_conf):

--- a/integration_tests/src/main/python/csv_test.py
+++ b/integration_tests/src/main/python/csv_test.py
@@ -15,6 +15,7 @@
 import pytest
 
 from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_fallback_collect, assert_gpu_fallback_write
+from conftest import get_non_gpu_allowed
 from datetime import datetime, timezone
 from data_gen import *
 from marks import *
@@ -288,7 +289,8 @@ def test_csv_fallback(spark_tmp_path, read_func, disable_conf):
             lambda spark : gen_df(spark, gen).write.csv(data_path))
     assert_gpu_fallback_collect(
             lambda spark : reader(spark).select(f.col('*'), f.col('_c2') + f.col('_c3')),
-            'FileSourceScanExec',
+            # TODO add support for lists
+            cpu_fallback_class_name=get_non_gpu_allowed()[0],
             conf=updated_conf)
 
 csv_supported_date_formats = ['yyyy-MM-dd', 'yyyy/MM/dd', 'yyyy-MM', 'yyyy/MM',

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala
@@ -392,17 +392,12 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
     }
   }
 
-  private def getBaseNameFromClass(planClassStr: String): String = {
-    val firstDotIndex = planClassStr.lastIndexOf(".")
-    if (firstDotIndex != -1) planClassStr.substring(firstDotIndex + 1) else planClassStr
-  }
-
   def assertIsOnTheGpu(exp: Expression, conf: RapidsConf): Unit = {
     // There are no GpuAttributeReference or GpuSortOrder
     if (!exp.isInstanceOf[AttributeReference] &&
         !exp.isInstanceOf[SortOrder] &&
         !exp.isInstanceOf[GpuExpression] &&
-      !conf.testingAllowedNonGpu.contains(getBaseNameFromClass(exp.getClass.toString))) {
+      !conf.testingAllowedNonGpu.contains(PlanUtils.getBaseNameFromClass(exp.getClass.toString))) {
       throw new IllegalArgumentException(s"The expression $exp is not columnar ${exp.getClass}")
     }
     exp.children.foreach(subExp => assertIsOnTheGpu(subExp, conf))
@@ -438,9 +433,9 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
         // Ignored for now, we don't force it to the GPU if
         // children are not on the gpu
       }
-      case other =>
-        if (!plan.supportsColumnar &&
-          !conf.testingAllowedNonGpu.contains(getBaseNameFromClass(other.getClass.toString))) {
+      case _ =>
+        if (!plan.supportsColumnar && !conf.testingAllowedNonGpu.exists(nonGpuClass =>
+          PlanUtils.sameClass(plan, nonGpuClass))) {
           throw new IllegalArgumentException(s"Part of the plan is not columnar " +
             s"${plan.getClass}\n${plan}")
         }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/PlanUtils.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/PlanUtils.scala
@@ -1,0 +1,47 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.util.Try
+
+import org.apache.spark.sql.execution.SparkPlan
+
+object PlanUtils {
+  def getBaseNameFromClass(planClassStr: String): String = {
+    val firstDotIndex = planClassStr.lastIndexOf(".")
+    if (firstDotIndex != -1) planClassStr.substring(firstDotIndex + 1) else planClassStr
+  }
+
+  /**
+   * Determines if plan is either fallbackCpuClass or a subclass thereof
+   *
+   * Useful subclass expression are LeafLike
+   *
+   * @param plan
+   * @param fallbackCpuClass
+   * @return
+   */
+  def sameClass(plan: SparkPlan, fallbackCpuClass: String): Boolean = {
+    val planClass = plan.getClass
+    val execNameWithoutPackage = getBaseNameFromClass(planClass.getName)
+    execNameWithoutPackage == fallbackCpuClass ||
+      plan.getClass.getName == fallbackCpuClass ||
+      Try(java.lang.Class.forName(fallbackCpuClass))
+        .map(_.isAssignableFrom(planClass))
+        .getOrElse(false)
+  }
+}

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala
@@ -318,28 +318,16 @@ object ExecutionPlanCaptureCallback {
         s"Could not find $fallbackCpuClass in the GPU plan\n$executedPlan")
   }
 
-  private def getBaseNameFromClass(planClassStr: String): String = {
-    val firstDotIndex = planClassStr.lastIndexOf(".")
-    if (firstDotIndex != -1) planClassStr.substring(firstDotIndex + 1) else planClassStr
-  }
-
   private def didFallBack(exp: Expression, fallbackCpuClass: String): Boolean = {
-    if (!exp.isInstanceOf[GpuExpression] &&
-      getBaseNameFromClass(exp.getClass.getName) == fallbackCpuClass) {
-      true
-    } else {
+    !exp.isInstanceOf[GpuExpression] &&
+      PlanUtils.getBaseNameFromClass(exp.getClass.getName) == fallbackCpuClass ||
       exp.children.exists(didFallBack(_, fallbackCpuClass))
-    }
   }
 
   private def didFallBack(plan: SparkPlan, fallbackCpuClass: String): Boolean = {
     val executedPlan = ExecutionPlanCaptureCallback.extractExecutedPlan(Some(plan))
-    if (!executedPlan.isInstanceOf[GpuExec] &&
-      getBaseNameFromClass(executedPlan.getClass.getName) == fallbackCpuClass) {
-      true
-    } else {
+    !executedPlan.isInstanceOf[GpuExec] && PlanUtils.sameClass(executedPlan, fallbackCpuClass) ||
       executedPlan.expressions.exists(didFallBack(_, fallbackCpuClass))
-    }
   }
 }
 

